component/common/loki/client/internal/marker_handler.go

```diff
@@ -0,0 +1,121 @@
+package internal
+
+import (
+	"github.com/grafana/agent/component/common/loki/wal"
+)
+
+type MarkerHandler interface {
+	wal.Marker
+
+	UpdateReceivedData(segmentId, dataCount int) // Data queued for sending
+	UpdateSentData(segmentId, dataCount int)     // Data which was sent or given up on sending
+
+	Stop()
+}
+
+type markerHandler struct {
+	markerFileHandler MarkerFileHandler
+	lastMarkedSegment int
+	dataIOUpdate      chan data
+	quit              chan struct{}
+}
+
+// TODO: Rename this struct
+type data struct {
+	segmentId int
+	dataCount int
+}
+
+var (
+	_ MarkerHandler = (*markerHandler)(nil)
+)
+
+func NewMarkerHandler(mfh MarkerFileHandler) MarkerHandler {
+	mh := &markerHandler{
+		lastMarkedSegment: -1, // Segment ID last marked on disk.
+		markerFileHandler: mfh,
+		//TODO: What is a good size for the channel?
+		dataIOUpdate: make(chan data, 100),
+		quit:         make(chan struct{}),
+	}
+
+	// Load the last marked segment from disk (if it exists).
+	if lastSegment := mh.markerFileHandler.LastMarkedSegment(); lastSegment >= 0 {
+		mh.lastMarkedSegment = lastSegment
+	}
+
+	//TODO: Should this be in a separate Start() function?
+	go mh.updatePendingData()
+
+	return mh
+}
+
+func (mh *markerHandler) LastMarkedSegment() int {
+	return mh.markerFileHandler.LastMarkedSegment()
+}
+
+func (mh *markerHandler) UpdateReceivedData(segmentId, dataCount int) {
+	mh.dataIOUpdate <- data{
+		segmentId: segmentId,
+		dataCount: dataCount,
+	}
+}
+
+func (mh *markerHandler) UpdateSentData(segmentId, dataCount int) {
+	mh.dataIOUpdate <- data{
+		segmentId: segmentId,
+		dataCount: -1 * dataCount,
+	}
+}
+
+func (mh *markerHandler) Stop() {
+	// Firstly stop the Marker Handler, because it might want to use the Marker File Handler.
+	mh.quit <- struct{}{}
+}
+
+// updatePendingData updates a counter for how much data is yet to be sent from each segment.
+// "dataCount" will be added to the segment with ID "dataSegment".
+func (mh *markerHandler) updatePendingData() {
+	batchSegmentCount := make(map[int]int)
+
+	for {
+		select {
+		case <-mh.quit:
+			return
+		case dataUpdate := <-mh.dataIOUpdate:
+			batchSegmentCount[dataUpdate.segmentId] += dataUpdate.dataCount
+		}
+
+		markableSegment := -1
+		for segment, count := range batchSegmentCount {
+			// TODO: If count is less than 0, then log an error and remove the entry from the map?
+			if count != 0 {
+				continue
+			}
+
+			// we know (segment, 0) is in the map
+
+			// TODO: Is it safe to assume that just because a segment is 0 inside the map,
+			//      all samples from it have been processed?
+			if segment > markableSegment {
+				markableSegment = segment
+			}
```

Comment:
I'm not sure if this is correct. For example:
* Let's have two batches - `batch_1` and `batch_2`.
* `batch_1` is at segment 3, whereas `batch_2` is at segment 5.
* `markableSegment` should be set at 3, whereas atm I think we will set it at 5

Reply:
@ptodev responding on the comment above